lib/publiccloud/provider.pm

```diff
@@ -269,23 +291,41 @@ sub terraform_apply {
     my $image                = $self->get_image_id();
     my $ssh_private_key_file = '/root/.ssh/id_rsa';
     my $name                 = get_var('PUBLIC_CLOUD_RESOURCE_NAME', 'openqa-vm');
+    my $cloud_name           = $self->conv_openqa_tf_name;
 
     record_info('WARNING', 'Terraform apply has been run previously.') if ($self->terraform_applied);
 
-    assert_script_run('cd ' . TERRAFORM_DIR);
     record_info('INFO', "Creating instance $instance_type from $image ...");
+    if (get_var('PUBLIC_CLOUD_SLES4SAP')) {
+        assert_script_run('cd ' . TERRAFORM_DIR . "/$cloud_name");
+        my $sap_media   = get_required_var('MEDIA');
+        my $sap_regcode = get_required_var('SCC_REGCODE_SLES4SAP');
+        my $sle_version = get_required_var('VERSION');
+        $sle_version =~ s/-/_/g;
+        my $cmd_tfvars = sprintf("sed -i 's|%MACHINE_TYPE%|%s|;s|%REGION%|%s|;s|%HANA_BUCKET%|%s|;s|%SLE_IMAGE%|%s|;s|%SCC_REGCODE_SLES4SAP%|%s|;s|%SLE_VERSION%|%s|' terraform.tfvars",
+            $instance_type, $self->region, $sap_media, $image, $sap_regcode, $sle_version);
+        record_info('TFM cmd_tfvars', $cmd_tfvars);
+        assert_script_run($cmd_tfvars);
+        upload_logs(TERRAFORM_DIR . "/$cloud_name/terraform.tfvars", failok => 1);
+        assert_script_run('terraform workspace new qashapopenqa -no-color', 120);
```

Comment:
Is `terraform workspace` the reason for `PUBLIC_CLOUD_ALWAYS_CLEANUP` ?

Reply:
> `120` fixed timeout?

Yes, but in fact it's a little bit high, 60s should be enough, especially if I will use `SCALE_TIMEOUT`.

> Is terraform workspace the reason for `PUBLIC_CLOUD_ALWAYS_CLEANUP`?

No the reason for `PUBLIC_CLOUD_ALWAYS_CLEANUP` is that I saw some cases during my tests where the VMs are not destroyed and that's very bad, mainly with SAP VMs, because they are big VMs and cost can be very expensive.

Reply:
But Why you don't use `TERRAFORM_TIMEOUT` ?

Reply:
I also didn't understand why we are using `workspace`. From what I have read so far. It is used to apply the same files twice and so getting more terraform states within one tf file. 
But if I'm not wrong, we apply each file only once or do I miss something?

Reply:
>But Why you don't use `TERRAFORM_TIMEOUT`?

Make sense yes, I will change for this.

As for `workspace`: it's mainly because `ha-sap-terraform-deployments` uses it for the VM name, and also because we have to create a workspace with our name (our openqa name in that case) to be able to quickly check who is doing what on the Webui of each PC provider. This is something asked by our managers. And as we (in our team) can manually deployed the same VM we can have trouble/failure if we all use the `default` workspace.

Reply:
>What is the value of ${terraform.workspace} if you don't use a workspace?

You always have a default one, so default value is `default` :-)

>say it's not supported by google and aws. Is this true?

We use workspace with GCP and AWS without any issue. And again needed also for GCP and AWS for tne VM name.
But what is weird also in the docs is that they are using `aws_instance` in their examples.

I will check for the tag also later, the main issue is that our `ha-sap-terraform-deployments` tool is now mainly developed by another team, we try to participate and we have regular discussion with SHAP devs but the tool is now customers focused. QA part is "just" a module in it.